driver-core/src/main/com/mongodb/internal/connection/ServerAddressWithResolver.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.MongoSocketException;
+import com.mongodb.ServerAddress;
+import com.mongodb.lang.Nullable;
+import com.mongodb.spi.dns.InetAddressResolver;
+import com.mongodb.spi.dns.InetAddressResolverProvider;
+
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+final class ServerAddressWithResolver extends ServerAddress {
```

Comment:
The important bit.  Why do we extend ServerAddress?  It's because the `StreamFactory` interface is part of the API, and defines the method `Stream create(ServerAddress serverAddress)`, and third-party implementations can assume they are allowed to call `ServerAddress#getSocketAddress` to resolve the IP address.  By extending ServerAddress, we can inject the InetAddressResolver behavior into all Stream implementations (including our own).